### PR TITLE
Adding ESFM system/insedit docs

### PR DESCRIPTION
**Pull request by Kagamiin**
Adding documentation for the ESFM system, including:

- Instrument editor
- Macros
- System documentation
- Effect list

### CODE REVIEW

**Comment by LTVA1** [comment]
Thanks for the docs. Will merge when ESFM is added to master branch (when the testing is done since I haven't tested it extnesively yet)

**Comment by Kagamiin** [comment]
@LTVA1 can you add @ElectricKeet as a reviewer please?

File: doc/4-instrument/fm-esfm.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+## FM
+
+ESFM is four-operator, meaning it takes four oscillators to produce a single sound.
+
+unlike most four-operator FM synthesizers, however, ESFM does not have an algorithm selection. instead, it uses a fixed operator arrangement, but allows you to independently control the output and modulation input levels of each operator. this allows it to reproduce a few common four-operator algorithms, as well as unique combinations where operators act as modulators and carriers at the same time.
+
+these apply to the instrument as a whole:
+
+- **OP4 Noise Mode**: enables special noise effects for operator 4 (0 to 3).
```
Comment (ElectricKeet) [inline doc/4-instrument/fm-esfm.md:20]:
```suggestion
- **OP4 Noise Mode**: enables special noise effects for operator 4. range is 0 to 3.
```

File: doc/4-instrument/fm-esfm.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  - these are normally used for the drum channels in ESFM's OPL3 compatibility mode.
+    - however, in ESFM, operator 4 can be modulated by operator 3, whereas in OPL3 drum mode the rhythm channels cannot be modulated.
+  - note: usage of noise mode 3 ("Top") is discouraged for now as it's not properly emulated yet, and results may change when the emulation gets fixed in the future.
+- **operator routing preview**: shows how operators are connected with each other and with the audio output (at the bottom).
+  - left-click pops up a small "operators changes with volume?" dialog where each operator can be toggled to scale with volume level.
+  - right-click to switch to a preview display of the waveform generated on a new note:
```
Comment (ElectricKeet) [inline doc/4-instrument/fm-esfm.md:26]:
```suggestion
  - right-click switches to a preview display of the waveform generated on a new note:
```

File: doc/4-instrument/fm-esfm.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+| OPL3 algorithm **3**           |  7  |  0  |  0  |  7  |  7  |  0  |  7  |
+| OPL3 algorithm **1** (variant) |  0  |  7  |  0  |  7  |  7  |  0  |  7  |
+
+### fixed frequency mode
+
+each operator has a Fixed Frequency mode. once enabled, the operator runs at the specified frequency regardless of the note.
+
+## macros
+
+these macros allow you to control several parameters of FM per tick.
```
Comment (ElectricKeet) [inline doc/4-instrument/fm-esfm.md:93]:
```suggestion
these macros allow you to control several parameters of FM on each tick.
```

Slightly clearer language (I hope?)

File: doc/4-instrument/fm-esfm.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+### envelope delay macro tricks
+
+due to a quirk in how the envelope delay feature works, the **Envelope Delay** macro can control the operator's key-on status for a limited amount of time after a note is played. a value of 0 represents key-on, while a value of 7 represents key-off.
+
+note that the macro cannot exceed 659.1 ms in length - anything beyond that will be treated as a value of 0.
+
+### operator arpeggio and pitch
+
+among the available macros are **Op. arpeggio** and **Op. pitch**. these work like the **Arpeggio** and **Pitch** macros featured below, but are applied individually over the operator, overriding the **Arpeggio**/**Pitch** macros respectively.
```
Comment (ElectricKeet) [inline doc/4-instrument/fm-esfm.md:107]:
```suggestion
among the available macros are **Op. arpeggio** and **Op. pitch**. these work like the **Arpeggio** and **Pitch** macros featured below, but are applied to the individual operator, overriding the **Arpeggio**/**Pitch** macros respectively.
```

File: doc/4-instrument/fm-esfm.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+## Macros
+
+- **Volume**: volume sequence.
+- **OP4 Noise Mode**: controls special noise effects for operator 4 (as described before).
+- **Arpeggio**: pitch sequence.
+- **Panning**: toggles output on left and right channels.
+  - note that each operator also has its own pan controls, which get masked by this global pan control.
```
Comment (ElectricKeet) [inline doc/4-instrument/fm-esfm.md:115]: Sorry, this seems a little unclear. Masked how?

File: doc/4-instrument/fm-esfm.md (hunk 4 above)
Comment (LTVA1) [reply doc/4-instrument/fm-esfm.md:107]: I would also add that despite them overriding main instrument pitch/arp macros, they still obey tune/detune settings set in FM tab

File: doc/4-instrument/fm-esfm.md (hunk 5 above)
Comment (Kagamiin) [reply doc/4-instrument/fm-esfm.md:115]: For instance, if an operator is set to only output on the right channel, and you set the global pan to only output on the left channel, it'll mask all of the operators' pan controls with the global pan setting - hence that operator that was set to output on the right channel won't output on any of the channels.

File: doc/4-instrument/fm-esfm.md (hunk 3 above)
Comment (Kagamiin) [reply doc/4-instrument/fm-esfm.md:93]: This was copypasted from the OPL page, but I totally agree.

File: doc/4-instrument/fm-esfm.md (hunk 5 above)
Comment (ElectricKeet) [reply doc/4-instrument/fm-esfm.md:125]: Now that you explain it, it seems a lot more obvious. I think this language will work fine!

**Pull request merged**